crates/ethrpc/src/current_block/mod.rs

```diff
@@ -16,6 +16,136 @@ use {
     },
 };
 
+/// Number of blocks that can be buffered in a [`broadcast::Receiver`].
+const MAX_SIZE: usize = 1_000;
+
+/// Internals of [`CurrentBlockStream`] that allow for only watching the most
+/// recent block as well as all blocks with the same handle.
+#[derive(Debug)]
+struct BlockStreamInner {
+    watch_receiver: watch::Receiver<BlockInfo>,
+    watch_sender: watch::Sender<BlockInfo>,
+    broadcast_sender: broadcast::Sender<BlockInfo>,
+}
+
+impl BlockStreamInner {
+    fn publish_block(&self, value: BlockInfo) {
+        let _ = self.broadcast_sender.send(value);
+        let _ = self.watch_sender.send(value);
+    }
+
+    fn new(initial_block: BlockInfo) -> Self {
+        let (watch_sender, watch_receiver) = watch::channel(initial_block);
+        let (broadcast_sender, _broadcast_receiver) = broadcast::channel(MAX_SIZE);
+
+        Self {
+            watch_receiver,
+            watch_sender,
+            broadcast_sender,
+        }
+    }
+}
+
+/// Monitors the blockchain for new blocks and allows users to get notified
+/// about them. Can be considered the "heartbeat" of the backend.
+#[derive(Debug, Clone)]
+pub struct CurrentBlockStream(Arc<BlockStreamInner>);
+
+impl CurrentBlockStream {
+    /// Returns a reference to the current block. Keeping this reference
+    /// alive locks a mutex so it should be used as shortly as possible.
+    pub fn current(&self) -> watch::Ref<'_, BlockInfo> {
```

Comment:
Now that we wrapped `.borrow()` should we consider not returning a reference here but the copy of data, so that no accidental lock is kept at callsite?

Reply:
Fine with me. 👌